### ISSUE TITLE
run a relay server

### ISSUE THREAD

**Issue by saljam** [saljam/webwormhole#57]
We should look into running a TURN server. It doesn't have to be fast, and it can be capped to whatever space capacity I have.

The fact that we don't ask browsers for media permissions means they only try WebRTC in "[Mode 2](https://tools.ietf.org/id/draft-ietf-rtcweb-ip-handling-12.html#rfc.section.5.2)". Plus, unless I misunderstood or am holding it wrong, Firefox and Chrome seem to exclude IPv6 from that mode, even if it's on the default route.

All this means we have slimmer chances of poking through NATs without relaying.

It's seems like the standard way to use short-lived auth token is this: https://tools.ietf.org/html/draft-uberti-behave-turn-rest-00

It also seems like everyone is using coturn.

**Comment by saljam** [comment]
It's seems like the standard way to use short-lived auth token is this: https://tools.ietf.org/html/draft-uberti-behave-turn-rest-00

**Comment by saljam** [comment]
Some more research:

TURN, by its design, invites abuse. Even with short-lived auth tokens (e.g. fetching these can be automated) it's still an open UDP(/TCP) proxy. If we only cared about DataChannels we'd be much better off relaying data through the signalling server, e.g. on the WebSocket. Alas, for this to work with other media types browsers only support TURN.

I can think of three different categories for abuse here:
1. Someone uses it to proxy into internal infrastructure.
2. Someone discovers how to get tokens and hardcodes it in their non-webwormhole application.
3. Someone uses it to masquerade their address while doing bad things on the internet, e.g. attack some unrelated targets.

1 is not applicable in this case. 

2 isn't great, but it's not the end of the world. If someone overuses it it'll run out of capacity (it's capped) and soon no one will be able to use it. Tragedy of the commons, etc. Plus, if it ever becomes a real problem we can try things like:
- disallow TURN on CORS requests.
- make relay sessions only valid within the slot. (probably implement turn in the signal server itself for this. pion/turn exists)

3 is bad, and I don't have the bandwidth to deal with the T&S issues that might fall out of it.

To deal with that, I found that one can configure coturn to only relay traffic to itself. This means relay only works if both peers are using it. I don't know if this is a common setup, but the fact that coturn defaults to allow for all remote peers and has config for a blacklist suggests it isn't.

If anyone is aware of a catch to doing it this way let me know.
